### ISSUE TITLE
Xamarin fails to compile and link project using Deedle and Fsharp.core libs

### ISSUE THREAD

**Issue by rolfbjarne** [mono/mono#17709]
_From @rhy-ama on Mon, 04 Nov 2019 18:03:32 GMT_

### Steps to Reproduce

1.  Unpack the provided Xamarin template project.
2.  The only XAML and class modified is the AboutViewModel xaml and cs.
3.  Set target to iPhone and try to compile

### Expected Behavior

Compilation without issues.

### Actual Behavior

Compilation fails and "Compile to native" stage. If Deedle and FSharp.core nugets are removed from the solution and relevant lines inside AboutViewModel.cs are commented/uncommented, code compiles and runs without issues.

### Environment

Visual studio on dev PC:

```
Microsoft Visual Studio 2019
Version 16.3.7
VisualStudio.16.Release/16.3.7+29424.173
Microsoft .NET Framework
Version 4.8.03752

Mono Debugging for Visual Studio   16.3.7 (9d260c5)
Support for debugging Mono processes with Visual Studio.

Visual F# Tools 10.4 for F# 4.6   16.3.0-beta.19455.1+0422ff293bb2cc722fe5021b85ef50378a9af823
Microsoft Visual F# Tools 10.4 for F# 4.6

Visual Studio Tools for CMake   1.0
Visual Studio Tools for CMake

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   16.3.0.277 (d16-3@c0fcab7)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   16.3.0.246 (remotes/origin/d16-3@bd2f86892)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   16.3.565 (27e9746)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   10.0.3.0 (d16-3/4d45b41)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: mono/mono/2019-06@5608fe0abb3
    Java.Interop: xamarin/java.interop/d16-3@5836f58
    LibZipSharp: grendello/LibZipSharp/d16-3@71f4a94
    LibZip: nih-at/libzip/rel-1-5-1@b95cf3fd
    ProGuard: xamarin/proguard/master@905836d
    SQLite: xamarin/sqlite/3.27.1@8212a2d
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-3@cb41333


Xamarin.iOS and Xamarin.Mac SDK   13.4.0.2 (e37549b)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.
```
Mac Catalina, XCode 11.2:
```
Version 8.3.6 (build 4)
	Xamarin.Mac 5.16.1.24 (d16-3 / 08809f5b)
	Package version: 604000208

=== Mono Framework MDK ===
Runtime:
	Mono 6.4.0.208 (2019-06/07c23f2ca43) (64-bit)
	Package version: 604000208

=== .NET Core SDK ===
SDK: /usr/local/share/dotnet/sdk/3.0.100/Sdks
SDK Versions:
	3.0.100
	2.1.701
	2.1.700
	2.1.505
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/6.4.0/lib/mono/msbuild/Current/bin/Sdks

=== .NET Core Runtime ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	3.0.0
	2.1.13
	2.1.12
	2.1.11
	2.1.9

=== Xamarin.Profiler ===
Version: 1.6.12.29
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Apple Developer Tools ===
Xcode 11.2 (15526)
Build 11B52

=== Xamarin.Mac ===
Version: 6.4.0.2 (Visual Studio Community)

=== Xamarin.iOS ===
Version: 13.4.0.2 (Visual Studio Community)
Hash: e37549bc
Branch: xcode11.1
Build date: 2019-10-07 22:43:23-0400

=== Xamarin.Android ===
Version: 10.0.3.0 (Visual Studio Community)
Commit: xamarin-android/d16-3/4d45b41
Android SDK: /Users/rhy-dev-mac02/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		8.1 (API level 27)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 28.0.2
SDK Build Tools Version: 28.0.3

Build Information: 
Mono: mono/mono/2019-06@5608fe0abb3
Java.Interop: xamarin/java.interop/d16-3@5836f58
LibZipSharp: grendello/LibZipSharp/d16-3@71f4a94
LibZip: nih-at/libzip/rel-1-5-1@b95cf3fd
ProGuard: xamarin/proguard/master@905836d
SQLite: xamarin/sqlite/3.27.1@8212a2d
Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-3@cb41333

=== Operating System ===

Mac OS X 10.15.1
Darwin 19.0.0 Darwin Kernel Version 19.0.0
    Thu Oct 17 16:17:15 PDT 2019
```

### Build Logs

[Build.txt](https://github.com/xamarin/xamarin-macios/files/3805388/Build.txt)

### Example Project (If Possible)

[DependencyTest.zip](https://github.com/xamarin/xamarin-macios/files/3805395/DependencyTest.zip)

_Copied from original issue xamarin/xamarin-macios#7361_

**Comment by rolfbjarne** [comment]
_From @chamons on Mon, 04 Nov 2019 21:37:31 GMT_

FYI @rolfbjarne 

Once I correctly swap it to 64-bit I hang on

```
  778   ??  R     11:20.60 /Library/Frameworks/Xamarin.iOS.framework/Versions/13.4.0.2/bin/arm64-darwin-mono-sgen --debug -O=gsharedvt -O=-float32 --aot=mtriple=arm64-ios,data-outfile=/Users/donblas/Downloads/DependencyTest/DependencyTest/DependencyTest.iOS/obj/iPhone/Debug/mtouch-cache/arm64/Deedle.aotdata.arm64,static,asmonly,direct-icalls,full,soft-debug,dwarfdebug,no-direct-calls,outfile=/Users/donblas/Downloads/DependencyTest/DependencyTest/DependencyTest.iOS/obj/iPhone/Debug/mtouch-cache/arm64/Deedle.dll.s /Users/donblas/Downloads/DependencyTest/DependencyTest/DependencyTest.iOS/obj/iPhone/Debug/mtouch-cache/3-Build/Deedle.dll
```
For a _long_ time.

**Comment by rolfbjarne** [comment]
_From @rhy-ama on Mon, 04 Nov 2019 21:42:04 GMT_

@chamons ....no worries. takes around an hour to drop and pop the error :) 

What do you mean by swapping to 64-bit? installing xamarin.ios-13.6.0.12.pkg from [here](https://github.com/xamarin/xamarin-macios/issues/7349) ?

Do you get it to compile on the mac ? i tried to compile remotely.

**Comment by rolfbjarne** [comment]
_From @rolfbjarne on Tue, 05 Nov 2019 11:46:55 GMT_

The problem is that the 2.3MB Deedle.dll turns into a 11GB Deedle.dll.s when AOT-compiled, then a 3.5GB Deedle.dll.o when the native code is compiled, and finally the native linker just gives up with a SIGBUS.

I think this is due to a generic explosion in the AOT compiler, but of a magnitude I've never seen before.

Moving to the mono repo, since if this is a bug, it's a bug in the AOT compiler.

**Comment by rhy-ama** [comment]
cc @FoggyFinder @zyzhu

@rolfbjarne thank you! obviously this event if it compiled and linked, 3.5GB file wont be accepted by apple. the sample code uses just one deedle object (SeriesBuilder). Could it be due to deedle dependencies to Fsharp.core ?

**Comment by rolfbjarne** [comment]
@rhy-ama I don't think it's the FSharp dependency, it's just something in Deedle.dll that makes the AOT compiler produce a lot of code.

In any case if you need a workaround, you can go to the project's iOS Build options, and select "Link All" for "Linker behavior", and the size issue should be fixed.

**Comment by rhy-ama** [comment]
@rolfbjarne thank you. no need for workaround - this is toy project to demo compile issue. link all could work in some real life situations. other threads mention dynamic code generation issues.

**Comment by zyzhu** [comment]
@rhy-ama A new Deedle 2.1.0 with explicit FSharp.Core dependency is published.

**Comment by marek-safar** [comment]
@vargaz any input on this?

**Comment by rhy-ama** [comment]
@zyzhu i've upgraded Deedle to 2.10 release (without updating yet to be release but available xamarin.ios-13.6.0.12.pkg ) - and "Compiling to native" been running the past half hour. Assume it does not make any difference (i will update if it does).

**Comment by vargaz** [comment]
I can repro on desktop with:
nuget install Deedle
mono --aot=full Deedle.dll

Deedle is written in f#, and it looks like the generics recursion detector in the AOT compiler is not good enough to prevent an explosion of generic instances.

**Comment by rhy-ama** [comment]
@vargaz is F# + Xamarin supported combination? The high use of generics was raised previously both here and in CoreRT and I thought it was addressed and supported. 

@marek-safar From [this (reddit)](https://www.reddit.com/r/dotnet/comments/dqdj49/eli5_what_is_microsoft_going_to_do_to_consolidate/) discussion one would infer mono AOT would take over "AOT toolchain from Mono will replace .NET Native, as discussed at BUILD". Is F# and .netcore AOT something that should be fully supported down the road?

**Comment by vargaz** [comment]
Not sure how to fix this. We could dial back the amount of generic instances we generate in AOT, but that could break existing apps.

**Comment by rhy-ama** [comment]
@vargaz any guidelines for Deedle developers to identify and "improve" the code?

cc @FoggyFinder @zyzhu

**Comment by charlesroddie** [comment]
> the sample code uses just one deedle object (SeriesBuilder)
@rhy-ama can you link to the definition of SeriesBuilder? It did not come up in a [search](https://github.com/fslaborg/Deedle/search?q=seriesbuilder&unscoped_q=seriesbuilder).

**Comment by FoggyFinder** [comment]
Definition of SeriesBuilder is here

[SeriesExtensions.fs#L115](https://github.com/fslaborg/Deedle/blob/4f1be1fbeb85ad11ed9a096cbf027c5708ba31ee/src/Deedle/SeriesExtensions.fs#L115)

**Comment by FoggyFinder** [comment]
> any guidelines for Deedle developers to identify and "improve" the code?

Exactly! I would like to help investigate F# related problems,  provide real MCVE and so on but I wasn't able even run `mono --aot` in Windows.

**Comment by charlesroddie** [comment]
> @rhy-ama any guidelines for Deedle developers to identify and "improve" the code?

Could do with more type annotations.

> Definition of SeriesBuilder is here

Looks straightforward apart from createGetterAndSetterFromFunc; probably try to make sure you're not using that aspect @rhy-ama .

You can try to verify the minimal repro by copy-pasting only the SeriesBuilder code in the library without referencing Deedle. Assuming it fails, you can try removing `interface System.Dynamic.IDynamicMetaObjectProvider` and seeing if it still fails.

**Comment by rhy-ama** [comment]
@charlesroddie thank you. Deedle maintainers are @FoggyFinder and @zyzhu, i was simply evaluating if the library was portable enough to use in a xamarin project. It seems there are still issues for it to be true. 

I can check what you suggest, but the best IMHO is for Deedle guys to be aware of the issue and have the capacity to reproduce and investigate it. I've seen @FoggyFinder conversation with CoreRT group, and this is along the same lines.

@FoggyFinder you would have to install [mono](https://www.mono-project.com/docs/getting-started/install/windows/) on windows or the platform of your choosing. Downloads are [here](https://www.mono-project.com/download/stable/) and i would suggest 64bit tooling.

**Comment by FoggyFinder** [comment]
> Deedle maintainers are @FoggyFinder and @zyzhu, i was simply evaluating if the library was portable enough to use in a xamarin project.

I'm just a usual contributor not maintainer :-)

> you would have to install mono on windows or the platform of your choosing. Downloads are here and i would suggest 64bit tooling.

Sadly, but it is not that simple. Default installation works fine until I was needed `aot` step.

`x64` version after installation gave me an error like: unable to find "clang.exe". Okay, I had installed it. Then it was unable to find "link.exe". Hrr, I had installed C++ for classical app in VS too. 
 
Finally some files `somename.exe.dll`, `somename.exe.dll.lib`, `somename.exe.dll.exp` were generated via aot but I didn't find any options to open any of them. When I tried to open them with mono I got ''Cannot open assembly '...': File does not contain a valid CIL image."
 
x32 version asks for "as". I gave up here. I had feeling that even if I'll install gnu assembler I got other errors or something else.
